File: LeetCodeRecord/ACFirst/654-ConstructMaximumBinaryTree.java
package ACFirst;


class ConstructMaximumBinaryTree {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }

    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return recur(nums, 0, nums.length - 1);
    } 

    public TreeNode recur(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }

        if (left == right) {
            return new TreeNode(nums[left]);
        }

        int max = nums[left];
        int maxIdx = left;
        for (int i = left + 1; i <= right; i++) {
            if (nums[i] > max) {
                max = nums[i];
                maxIdx = i;
            }
        }

        TreeNode curr = new TreeNode(max);
        curr.left = recur(nums, left, maxIdx - 1);
        curr.right = recur(nums, maxIdx + 1, right);

        return curr;
    } 
}

File: Top200/654-ConstructMaximumBinaryTree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructMaximumBinaryTree(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	return constructRecur(nums, 0, len(nums)-1)
}

func constructRecur(nums []int, left, right int) *TreeNode {
	if left > right {
		return nil
	}

	index := left
	max := nums[left]
	for i := left + 1; i <= right; i++ {
		if nums[i] > max {
			index = i
			max = nums[i]
		}
	}

	root := &TreeNode{
		Val: nums[index],
	}

	root.Left = constructRecur(nums, left, index-1)
	root.Right = constructRecur(nums, index+1, right)

	return root
}
